### ISSUE TITLE
is_positive_definite gives wrong result

### ISSUE THREAD

**Issue by rikardn** [sympy/sympy#18955]
The is_positive_definite does not always give a result. In the example below with a 4x4 matrix, which numpy can demonstrate is posdef, sympy returns neither False nor True but None. None is interpreted as False in an if statement and the result thus becomes wrong.

```python
import numpy as np
import sympy

M = [[0.0228202735623867, 0.00518748979085398, -0.0743036351048907, -0.00709135324903921],
    [0.00518748979085398, 0.0349045359786350, 0.0830317991056637, 0.00233147902806909],
    [-0.0743036351048907, 0.0830317991056637, 1.15859676366277, 0.340359081555988],
    [-0.00709135324903921, 0.00233147902806909, 0.340359081555988, 0.928147644848199]]

M_sympy = sympy.Matrix(M)
print(M_sympy.is_positive_definite)

M_numpy = np.array(M)
print(np.linalg.cholesky(M_numpy))
```

**Comment by abhinav28071999** [comment]
Just a  question, if a matrix is positive_definite the its eigen values should be positive. If the eigenvalues are complex then does it also require the imaginary part to be positive or not?

**Comment by abhinav28071999** [comment]
The problem is when `is_definite` is used then sympy calculate the eigenvalues of the matrix, and in this case it seems to be too complex.   When i do evalf of the eigenvals it gives
0.0351585880190815 - 2.83824441184299e-30*I ... , But this imaginary part should not be here.
Instead of using eigenvalues we can use pivots of the matrix to determine if it's positive definite or not. Calculating  eigenvals seems too complex.

**Comment by oscarbenjamin** [comment]
For the matrix to be positive definite the eigenvalues have to be real and positive.

**Comment by sylee957** [comment]
I don't think that it is a completely wrong result because at least it returns fuzzy `None`, but the problem is that the matrix returns complicated expressions that makes sympy undecidable whether the result is positive or negative.
(I'd claim that testing `real` is redundant since you should assume all the eigenvalues to be real if the matrix is symmetric)

I think that there are different ways to test positive-definiteness instead of relying on roots, if it arrives on this result.
For example, positive-definiteness can be computed only using the determinants
https://en.wikipedia.org/wiki/Sylvester%27s_criterion

Or some root isolation techniques from the characteristic polynomial can be used to check this stuff too
https://en.wikipedia.org/wiki/Routh%E2%80%93Hurwitz_stability_criterion

**Comment by sylee957** [comment]
And usually, I'd use `x is False` instead of `not x` if you cannot expect that the problem can be decidable to True, False perfectly in sympy.

**Comment by rikardn** [comment]
Thanks @sylee957 for pointing out that None as a return value means that sympy couldn't determine the definiteness. I wasn't able to find that information in the documentation though.

Since this case is purely numeric you would expect a True or False result.

The sympy cholesky give:

```
M_sympy.cholesky()
Out[5]: 
Matrix([
[  0.151063806262078,                  0,                 0,                 0],
[ 0.0343397264984459,  0.183644545692615,                 0,                 0],
[ -0.491869210391684,  0.544107928099162, 0.787786777060309,                 0],
[-0.0469427682547369, 0.0214734439084148, 0.387903830103733, 0.880348528228648]])
```

But I don't know how good the symbolic cholesky is for more general cases

**Comment by sylee957** [comment]
I see determining positive-definiteness by pivoting is equivalent to 
- Testing positive-definiteness by evaluating all the principal minor determinants
- Testing positive when performing cholesky decomposition

Because division-free gaussian elimination may reveal the determinant of principal minors alongside the diagonals. (Though I'm not sure how to make this generalized for testing positive-semidefinite matrix as well.)

Regarding https://github.com/sympy/sympy/issues/18955#issuecomment-604276639, actually, for numeric cases, computing cholesky decomposition is equivalent to doing positive-definiteness and usually there is no waste of computing that way.

And I think that I had made the positive-definite tests using pivoting as a non-public API
So if you access via `M_sympy._eval_is_positive_definite(method='CH')` it will return `True`.

**Comment by sylee957** [comment]
I made a new framework about this using division free gaussian elimination.
And I think that it would be mostly faster than computing the eigenvalues because computing eigenvalues would already involve computing determinants. c15f945406b6c344cc9c9a9e706daa5fcb671ecd
Though there are some cases where computing eigenvalues can be economical, like triangular matrix. So I'm only going to push the changes after #18911 is done to clean up the eigenvalue code.

**Comment by sylee957** [comment]
This was alternatively fixed by #18911 but some arguments remain to add an alternative test for positive definite.
